perfkitbenchmarker/windows_virtual_machine.py

```diff
@@ -0,0 +1,306 @@
+      copy_to: True to copy to vm, False to copy from vm.
+
+    Raises:
+      RemoteCommandError: If there was a problem copying the file.
+    """
+    if ':' in remote_path:
+      remote_path = remote_path.split(':', 1)[1]
+
+    set_error_pref = '$ErrorActionPreference="Stop"'
+
+    password = self.password.replace("'", "''")
+    create_cred = (
+        '$pw = convertto-securestring -AsPlainText -Force \'%s\';'
+        '$cred = new-object -typename System.Management.Automation'
+        '.PSCredential -argumentlist %s,$pw' % (password, self.user_name))
+
+    psdrive_name = self.name
+    root = '\\\\%s\\c$' % self.ip_address
+    create_psdrive = (
+        'New-PSDrive -Name %s -PSProvider filesystem -Root '
+        '%s -Credential $cred' % (psdrive_name, root))
+
+    remote_path = '%s:%s' % (psdrive_name, remote_path)
+    if copy_to:
+      from_path, to_path = local_path, remote_path
+    else:
+      from_path, to_path = remote_path, local_path
+
+    copy_item = 'Copy-Item -Path %s -Destination %s' % (from_path, to_path)
+
+    delete_connection = 'net use %s /delete' % root
+
+    cmd = ';'.join([set_error_pref, create_cred, create_psdrive,
+                    copy_item, delete_connection])
+
+    stdout, stderr, retcode = vm_util.IssueCommand(
+        ['powershell', '-Command', cmd], timeout=None)
+
+    if retcode:
+      error_text = ('Got non-zero return code (%s) executing %s\n'
+                    'STDOUT: %sSTDERR: %s' %
+                    (retcode, cmd, stdout, stderr))
+      raise errors.VirtualMachine.RemoteCommandError(error_text)
+
+  @vm_util.Retry(log_errors=False, poll_interval=1)
+  def WaitForBootCompletion(self):
+    """Waits until VM is has booted."""
+    resp, _ = self.RemoteCommand('hostname', suppress_warning=True)
+    if self.bootable_time is None:
+      self.bootable_time = time.time()
+    if self.hostname is None:
+      self.hostname = resp[:-1]
+
+  def OnStartup(self):
+    stdout, _ = self.RemoteCommand('echo $env:TEMP')
+    self.temp_dir = os.path.join(stdout.strip(), 'pkb')
+    self.RemoteCommand('mkdir %s' % self.temp_dir)
+    self.DisableGuestFirewall()
+
+  def Install(self, package_name):
+    """Installs a PerfKit package on the VM."""
+    if ((self.is_static and not self.install_packages) or
+        not FLAGS.install_packages):
+      return
+    if package_name not in self._installed_packages:
+      package = windows_packages.PACKAGES[package_name]
+      package.Install(self)
+      self._installed_packages.add(package_name)
+
+  def Uninstall(self, package_name):
+    """Uninstalls a Perfkit package on the VM."""
+    package = windows_packages.PACKAGES[package_name]
+    if hasattr(package, 'Uninstall'):
+      package.Uninstall()
+
+  def PackageCleanup(self):
+    """Cleans up all installed packages.
+
+    Deletes the temp directory and uninstalls all PerfKit packages.
+    """
+    for package_name in self._installed_packages:
+      self.Uninstall(package_name)
```

Comment:
The only case where this could happen is with a static vm. With static vms, we don't try to detect if the package is already present on the system - we just have two modes - either we installed it (in which case we will uninstall it here), or we didn't install any packages (in which case we don't uninstall it here). There is a config option to set this on a per VM basis and a flag override.